lib/segment/src/index/sparse_index/sparse_vector_index.rs

```diff
@@ -101,19 +102,35 @@ impl<TInvertedIndex: InvertedIndex> SparseVectorIndex<TInvertedIndex> {
                 || (),
             )?;
             (config, inverted_index, indices_tracker)
-        } else if config_path.exists() {
-            // Load inverted index and config
-            migrate(path)?;
-            let loaded_config = SparseIndexConfig::load(&config_path)?;
-            let inverted_index = TInvertedIndex::open(path)?;
-            let indices_tracker =
-                IndicesTracker::open(path, || inverted_index.max_index().unwrap_or_default())?;
-            (loaded_config, inverted_index, indices_tracker)
         } else {
-            // Inverted index and config are not presented - initialize empty inverted index
-            let inverted_index = TInvertedIndex::from_ram_index(InvertedIndexRam::empty(), path)?;
-            let indices_tracker = Default::default();
-            (config, inverted_index, indices_tracker)
+            Self::try_load(path).or_else(|e| {
+                // Avoid noisy warning for newly created segments
+                if vector_storage.borrow().total_vector_count() != 0 {
+                    log::warn!("Failed to load, rebuilding: {}", e.to_string());
+                }
+
+                let (inverted_index, indices_tracker) = Self::build_inverted_index(
+                    id_tracker.clone(),
+                    vector_storage.clone(),
+                    path,
+                    stopped,
+                    || (),
+                )?;
+
+                remove_dir_all(path)?;
+                create_dir_all(path)?;
+
+                SparseVectorIndexVersion::save(path)?;
+                config.save(&config_path)?;
+                inverted_index.save(path)?;
+                indices_tracker.save(path)?;
```

Comment:
I am not sure if this order of saving will reliable once we assume that service can be killed at any moment.

Assume the following scenario:

- we migrate from the current version
- we do the save 

```
SparseVectorIndexVersion::save(path)?;
config.save(&config_path)?;
inverted_index.save(path)?;
```

but the service got killed with SIGKILL

during the

```
indices_tracker.save(path)?;
```

In this case next time we will load the index, it will already see the new version and won't try to apply migration, even though the `indices_tracker` was only partially saved.

It might be, that partially saved index will cause the error and re-building of the index on it's own, but there are no explicit guarantees about that.

Our suggestion would be to change the order of saving and save version file always as a last step. In this way we can guarantee that the migration is completed if we see the newest version in the storage.

Reply:
> It might be, that partially saved index will cause the error and re-building of the index on it's own, but there are no explicit guarantees about that.

It will fail during `try_load()` which requires all of the files to be present.
Note: previously, the missing indices tracker file was optional, falling back to trivial mapping (`indices_tracker.remap_index(i) == Some(i)`). In this PR this file is required to be present.

I think "all files are required" is easier to reason than considering the order of saves.

Perhaps, except, for the version file, to address your review comment below. It still should not be the last file to save. Otherwise, in following scenario the resulting files would be implicitly incorrectly treated as the old format due to the missing version file.
```rs
config.save(); inverted_index.save(); indices_tracker.save();
SIGKILL
version.save(); // version not saved
```

I'll state this approach explicitly in the code comments.

Reply:
> It will fail during try_load() which requires all of the files to be present.

This is only true if the `indices_tracker.save(path)` is atomic. But in current implementation it is only atomic _by chance_. 
Nobody would stop us if we would want to implement it in a way like `inverted_index.save` is implemented, and this issue will be very hard to find. 


> Otherwise, in following scenario the resulting files would be implicitly incorrectly treated as the old format due to the missing version file.

There are 2 options here:

1.  use different names for new version files (which also will make it easier to rollback to previous version in case of failure and don't delete old index before save)

2. Delete version file before saving index - easy, but not as elegant